src/operator.ts

```diff
@@ -0,0 +1,65 @@
+/**
+ * OperatorOptions define shared properties for all Operators.
+ */
+export interface OperatorOptions {
+  name: string;             // the name of the operator
+  index?: number;           // an index if the operation in on a specific object in a list
+  maxDepth?: number;        // the maximum depth to traverse nested objects
+}
+
+/**
+ * Common validation errors, which prevents each Operator from defining their own strings and bloating code.
+ */
+export enum OperatorValidationError {
+  MISSING = 'is missing or empty',
+  UNSUPPORTED = 'has unsupported value',
+  MALFORMED = 'is malformed'
+}
+
+/**
+ * An Operator takes a list of data objects emitted from a data layer and performs a transformation.
+ * Operators should be atomic - performing a specific and succinct transformation. The transformation should
+ * return a result that will be passed to the next operator, and it should not mutate the incoming object.
+ *
+ * While an Operator will most often transform a single object, the input is always a list of objects.
+ * This is because data emitted from the data layer could be a single value or the result of a function call, which
+ * emits a list of arguments.
+ *
+ * An Operator can choose not to pass information to the next operator by returning null.
+ */
+export abstract class Operator {
+  readonly name: string;
```

Comment:
Would it make sense to hold the OperatorOptions instance instead of transferring its data to properties on Operator? This would give us some wiggle room for additions to options (both data and logic) in the future without changing the constructor API.

Reply:
Made changes to accommodate `options` as a member of the class.  Due to TS compiler checking, the `index` member needs to remain as it must hold a value to support `data[index]` usage in `handleData()`.  Trevor and I both agreed on the approach.